scarb/src/core/lockfile.rs

```diff
@@ -23,12 +25,44 @@ pub enum LockVersion {
 
 #[derive(Debug, Clone, Eq, PartialEq, Default, Serialize, Deserialize)]
 #[serde(rename_all = "kebab-case")]
+#[serde(from = "serdex::Lockfile", into = "serdex::Lockfile")]
 pub struct Lockfile {
     pub version: LockVersion,
-    #[serde(rename = "package")]
-    #[serde(default = "BTreeSet::new")]
-    #[serde(skip_serializing_if = "BTreeSet::is_empty")]
-    pub packages: BTreeSet<PackageLock>,
+    // Note that the current compilation model will not allow Scarb to emit lock files with multiple
+    // entries for the same package name. This structure allows this for the sake of future-proofing
+    // and being more defensive against lock files tinkered with by hand. Thus, we use `SmallVec`
+    // to optimise for the common case of a single entry only.
+    pub packages: BTreeMap<PackageName, SmallVec<[PackageLock; 1]>>,
```

Comment:
I think at this stage of future-proofing, it might be good idea to make these fields private